src/actions/pkg.cr

```diff
@@ -2,49 +2,172 @@ module DppmRestApi::Actions::Pkg
+            "received empty set from Prefix#clean_unused_packages; please report this strange bug",
+          ],
+        }
+      else
+        result
+      end
+    else
+      ctx.response.status_code = 404
+      {errors: ["no packages to clean"]}
+    end
+  end
+
   # List built packages
-  relative_get ALL_PKGS do |context|
-    if context.current_user? && Config.has_access? context, Access::Read
-      # TODO: list all built packages
-      next context
+  relative_get ALL_PKGS do |ctx|
+    if ctx.current_user? && Config.has_access? ctx, Access::Read
+      pfx = Prefix.new prefix
+      # Build JSON directly to the HTTP response IO
+      JSON.build ctx.response do |json|
+        json.array do
+          # build an array of all the package names
+          pfx.each_pkg { |pkg| json.string pkg.package }
+        end
+      end
+      next ctx
     end
-    deny_access! to: context
+    deny_access! to: ctx
   end
   # Clean unused built packages
-  relative_delete ALL_PKGS do |context|
-    if context.current_user? && Config.has_access? context, Access::Delete
-      # TODO: delete all unused built packages
-      next context
+  relative_delete ALL_PKGS do |ctx|
+    puts "in clean packages route"
+    if ctx.current_user? && Config.has_access? ctx, Access::Delete
+      clean_unused_packages(ctx).to_json ctx.response
+      ctx.response.flush
+      next ctx
     end
-    deny_access! to: context
+    deny_access! to: ctx
   end
   # Query information about a given package
-  relative_get ONE_PKG do |context|
-    if context.current_user? && Config.has_access? context, Access::Read
-      # TODO: Query information about the given package
-      next context
+  #
+  # Query parameter:
+  #   get=config_key      request a particular config key.
+  #
+  # TODO: allow specifying more than one config key like ?get=item1&get=item2
+  relative_get ONE_PKG do |ctx|
+    if ctx.current_user? && Config.has_access? ctx, Access::Read
+      package_name = URI.unescape ctx.params.url["id"]
+      # Iterate over the packages to find the relevant one.
+      if selected_pkg = find_package_by_name package_name, prefix
+        # The package whose name was the :id URL parameter was found and we can query it
+        if key = ctx.params.query["get"]?
+          # A key was specified by the &get query parameter
+          begin
+            {
+              package_name => {
+                key => selected_pkg.get_config key,
+                # This method call  ^^ raises an untyped error if the key is not found
+              },
+              "errors" => [] of Nil,
+            }.to_json ctx.response
+            ctx.response.flush
+            next ctx
+          rescue e : Exception
+            # catch the exception thown by the lack of a config value for the specified key
+            # otherwise, pass it on to the
+            msg = e.message || raise e
+            raise e unless msg.starts_with? "config key not found"
```

Comment:
No need to do this. `get_config` has a block variant that yields when the key isn't found https://github.com/DFabric/dppm/blob/master/src/prefix/app.cr#L197
I'm creating a custom error anyway. Not sure what option is the best. On one hand using an exception brings a standard message, on the other hand the block one will be faster and more elegant.

Reply:
You can also now catch `ConfigKeyError`